interfaces/repo.go

```diff
@@ -80,6 +80,112 @@ func (r *Repository) AddInterface(i Interface) error {
 	return nil
 }
 
+// QueryOptions describes options for QueryInterfaces.
+//
+// Names: return just this subset if non-empty.
+// Doc: return documentation.
+// Plugs: return information about plugs.
+// Slots: return information about slots.
+// Connected: only consider interfaces with at least one connection.
+type QueryOptions struct {
+	Names                        []string
+	Doc, Plugs, Slots, Connected bool
+}
+
+func (r *Repository) queryInterface(iface Interface, opts *QueryOptions) *InterfaceInfo {
+	// NOTE: QueryOptions.Connected is handled by QueryInterfaces
+	md := MetaDataOf(iface)
+	ifaceName := iface.Name()
+	ii := &InterfaceInfo{
+		Name:     ifaceName,
+		MetaData: MetaData{Summary: md.Summary},
+	}
+	if opts != nil && opts.Doc {
+		// Collect documentation URL
+		ii.MetaData.DocURL = md.DocURL
+	}
+	if opts != nil && opts.Plugs {
+		// Collect all plugs of this interface type.
+		for _, snapName := range sortedSnapNamesWithPlugs(r.plugs) {
+			for _, plugName := range sortedPlugNames(r.plugs[snapName]) {
+				plugInfo := r.plugs[snapName][plugName].PlugInfo
+				if plugInfo.Interface == ifaceName {
+					ii.Plugs = append(ii.Plugs, plugInfo)
+				}
+			}
+		}
+	}
+	if opts != nil && opts.Slots {
+		// Collect all slots of this interface type.
+		for _, snapName := range sortedSnapNamesWithSlots(r.slots) {
+			for _, slotName := range sortedSlotNames(r.slots[snapName]) {
+				slotInfo := r.slots[snapName][slotName].SlotInfo
+				if slotInfo.Interface == ifaceName {
+					ii.Slots = append(ii.Slots, slotInfo)
+				}
+			}
+		}
+	}
+	return ii
+}
+
+// QueryInterfaces returns information about interfaces in the system.
+//
+// If names is empty then all interfaces are considered. Query options decide
+// which data to return but can also skip interfaces without connections. See
+// the documentation of QueryOptions for details.
+func (r *Repository) QueryInterfaces(opts *QueryOptions) []*InterfaceInfo {
```

Comment:
This can be the Info method, taking InfoOptions in and returning []*interfaces.Info.